frontend_lib/src/mention.js

```diff
@@ -17,111 +17,101 @@ export const GROUP_MENTION_LIST = [
-      i++
-    }
-  })
+const wrapMentionsInText = (doc, text) => {
+  const match = text.match(MENTION_REGEX)
+
+  if (!match) {
+    return doc.createTextNode(text)
+  }
+
+  const fragment = doc.createDocumentFragment()
+
+  fragment.appendChild(doc.createTextNode(text.substring(0, match.index)))
+
+  const wrappedMention = doc.createElement(MENTION_TAG_NAME)
+  wrappedMention.className = MENTION_CLASS
+  wrappedMention.id = `${MENTION_ID_PREFIX}${uuidv4()}`
+  wrappedMention.textContent = match[0]
+  fragment.appendChild(wrappedMention)
+
+  const mentionEndIndex = match.index + match[0].length
+  fragment.appendChild(wrapMentionsInText(doc, text.substring(mentionEndIndex)))
+
+  return fragment
 }
 
-// Call the given callback for each mention found in the given content.
-// Mention detection assumes that the mentions have been wrapped using wrapMentionsInSpanTags().
-// callback with get the matching DOM Element as parameter.
-// document is a DOM Document
-const forEachMentionIn = (callback, document, mentions) => {
-  const elementHasMentionForUser = element => {
-    if (element.id === null || !element.id.startsWith(MENTION_ID_PREFIX)) return false
-    for (const mention of mentions) {
-      if (element.textContent.includes('@' + mention)) return true
+const isAWrappedMention = (node) => (
+  node.nodeName.toLowerCase() === MENTION_TAG_NAME &&
+  node.classList.contains(MENTION_CLASS) &&
+  (node.id || '').startsWith(MENTION_ID_PREFIX)
+)
 
-    }
-    return false
+export const wrapMentionsInSpanTags = (doc, node) => {
+  if (isAWrappedMention(node)) {
+    return node.cloneNode(true)
   }
 
-  const spans = document.getElementsByTagName(MENTION_TAG_NAME)
-  for (const element of spans) {
-    if (elementHasMentionForUser(element)) callback(element)
+  const resultingNode = node.cloneNode(false)
+
+  for (const child of node.childNodes) {
+    resultingNode.appendChild(
+      (child.nodeName === '#text')
+        ? wrapMentionsInText(doc, child.textContent)
+        : wrapMentionsInSpanTags(doc, child)
+    )
   }
+
+  return resultingNode
 }
 
-export const addClassToMentionsOfUser = (rawContent, username, userClassName = MENTION_ME_CLASS) => {
-  const addUserClass = element => element.classList.add(userClassName)
+const getDocumentFromHTMLString = (htmlString) => {
+  const doc = new DOMParser().parseFromString(htmlString, 'text/html')
 
-  const parser = new DOMParser()
-  const document = parser.parseFromString(rawContent, 'text/html')
-  if (document.documentElement.tagName === 'parsererror') {
-    throw new Error('Cannot parse string: ' + document.documentElement)
+  if (doc.documentElement.tagName === 'parsererror') {
+    throw new Error('Cannot parse string: ' + doc.documentElement.textContent)
+  }
+
+  return doc
+}
+
+const getMentions = function* (node) {
+  for (const candidate of node.querySelectorAll(MENTION_TAG_NAME)) {
+    if (candidate.id.startsWith(MENTION_ID_PREFIX)) {
```

Comment:
Shouldn't you test for candidate.id true-ness (as you do below)?

Reply:
BTW isn't `isAWrappedMention` applicable?

Reply:
Indeed, but isAWrappedMention checked the presence of the mention class, which made it not usable there. I removed this check in this function and used the function in getMentions.